File: z12-objective-c-interoperability/projects/final/BabyTrack/UIKit/FeedCell.m
/// Sample code from the book, Expert Swift,
/// published at raywenderlich.com, Copyright (c) 2021 Razeware LLC.
/// See LICENSE for details. Thank you for supporting our work!
/// Visit https://www.raywenderlich.com/books/expert-swift

#import "FeedCell.h"
#import "BabyTrack-Swift.h"

@interface FeedCell ()
@property (nonatomic, strong) IBOutlet UILabel *lblKindEmoji;
@property (nonatomic, strong) IBOutlet UILabel *lblKindTitle;
@property (nonatomic, strong) IBOutlet UILabel *lblDate;
@property (nonatomic, strong) IBOutlet UIImageView *attachment;
@end

@implementation FeedCell

- (void)configureWithFeedItem:(FeedItem *)feedItem {
  self.lblKindEmoji.text = [FeedItem emojiForKind:feedItem.kind];
  self.lblKindEmoji.backgroundColor = [FeedItem colorForKind:feedItem.kind];
  self.lblKindTitle.text = [FeedItem titleForKind:feedItem.kind];

  NSRelativeDateTimeFormatter *df = [NSRelativeDateTimeFormatter new];
  df.formattingContext = NSFormattingContextStandalone;
  self.lblDate.text = [df stringForObjectValue:feedItem.date];
  
  [self.attachment setHidden: feedItem.attachmentId == nil];
  
  [[NSURLSession.sharedSession dataTaskWithURL:feedItem.attachmentURL
                             completionHandler:^(NSData * _Nullable data,
                                                 NSURLResponse * _Nullable response,
                                                 NSError * _Nullable error) {

    __weak FeedCell *weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
      weakSelf.attachment.image = [UIImage imageWithData:data];
    });
  }] resume];
}

@end

File: z12-objective-c-interoperability/projects/starter/BabyTrack/SwiftUI/FeedCell.swift
/// Sample code from the book, Expert Swift,
/// published at raywenderlich.com, Copyright (c) 2021 Razeware LLC.
/// See LICENSE for details. Thank you for supporting our work!
/// Visit https://www.raywenderlich.com/books/expert-swift

import SwiftUI
import BabyKit

struct FeedCell: View {
  private let feedItem: FeedItem
  private let formatter = RelativeDateTimeFormatter()

  init(feedItem: FeedItem) {
    self.feedItem = feedItem
    formatter.formattingContext = .standalone
  }

  var body: some View {
    HStack(alignment: .top) {
      Text(feedItem.kind.emoji)
        .frame(width: 44, height: 44, alignment: .center)
        .background(Color(feedItem.kind.color))
        .cornerRadius(22)

      VStack(alignment: .leading) {
        Text(feedItem.kind.title)
          .font(.headline)

        Text(feedItem.date, formatter: formatter)
          .font(.subheadline)
          .foregroundColor(.gray)

        if let attachmentURL = feedItem.attachmentURL {
          Spacer(minLength: 4)

          URLImage(url: attachmentURL)
            .frame(minWidth: 0, minHeight: 140, idealHeight: 140)
            .clipped()
        }
      }

      Spacer()
    }
    .frame(minHeight: 64)
  }
}
